src/pages/studyView/geneSymbolSelection/GeneSelectionBox.tsx

```diff
@@ -0,0 +1,343 @@
+        }
+    }
+
+    @computed private get geneIds(): string[] {
+        try {
+            return this.oql.query.map(line => line.gene);
+        }
+        catch (e) {
+            return [];
+        }
+    }
+
+    @memoize
+    private async getGeneSuggestions(alias: string): Promise<GeneReplacement> {
+        return {
+            alias,
+            genes: await client.getAllGenesUsingGET({ alias })
+        };
+    }
+
+    private invokeGenesLater = debounceAsync(
+        async (geneIds: string[]): Promise<{ found: Gene[], suggestions: GeneReplacement[] }> => {
+            let [entrezIds, hugoIds] = _.partition(_.uniq(geneIds), isInteger);
+
+            let getEntrezResults = async () => {
+                let found: Gene[];
+                if (entrezIds.length)
+                    found = await client.fetchGenesUsingPOST({ geneIdType: "ENTREZ_GENE_ID", geneIds: entrezIds });
+                else
+                    found = [];
+                let missingIds = _.difference(entrezIds, found.map(gene => gene.entrezGeneId + ''));
+                let removals = missingIds.map(entrezId => ({ alias: entrezId, genes: [] }));
+                let replacements = found.map(gene => ({ alias: gene.entrezGeneId + '', genes: [gene] }));
+                let suggestions = [...removals, ...replacements];
+                return { found, suggestions };
+            };
+
+            let getHugoResults = async () => {
+                let found: Gene[];
+                if (hugoIds.length)
+                    found = await client.fetchGenesUsingPOST({ geneIdType: "HUGO_GENE_SYMBOL", geneIds: hugoIds });
+                else
+                    found = [];
+                let missingIds = _.difference(hugoIds, found.map(gene => gene.hugoGeneSymbol));
+                let suggestions = await Promise.all(missingIds.map(alias => this.getGeneSuggestions(alias)));
+                return { found, suggestions };
+            };
+
+            let [entrezResults, hugoResults] = await Promise.all([getEntrezResults(), getHugoResults()]);
+            return {
+                found: [...entrezResults.found, ...hugoResults.found],
+                suggestions: [...entrezResults.suggestions, ...hugoResults.suggestions]
+            };
+        },
+        500
+    );
+
+    private readonly genes = remoteData({
+        invoke: () => this.invokeGenesLater(this.geneIds),
+        default: { found: [], suggestions: [] },
+        onResult: (genes) => {
+            let status = _.isUndefined(this.oql.error) && genes.suggestions.length===0;
+            this.props.callback(this.oql.query, genes.found, this._geneQuery, status)
+        }
+    });
+
+    private validationStatus() {
+
+        if (this.oql.error)
+            return (
+                <span className={styles.errorMessage}>
+                    {`Cannot validate gene symbols because of invalid OQL. ${
+                        this._geneQueryErrorDisplayStatus === 'unfocused'
```

Comment:
I think its better practice to use constants/enums instead of the string values, like 'unfocused' and on line 275 'studyView'. That way the compiler can catch any typos which could otherwise become pernicious bugs, and also when programming the IDE can autocomplete